### DISCUSSION TITLE
Re:  Flowing-State-in-Blazor-8

### DISCUSSION THREAD

**Discussion by swegele** [MarimerLLC/csla#3537] [General]
@rockfordlhotka  
Appreciate your last 2 blog articles about the new Blazor deployment styles in .NET 8 and the subsequent issues with state.  
[Flowing-State-in-Blazor-8](https://blog.lhotka.net/2023/10/27/Flowing-State-in-Blazor-8)

For years (too many to admit) I've enjoyed checking in regularly on your blog and finding road signs that let me know what is coming in the near future that may be of concern to me and other LOB app devs.

I wanted to add a thought related to your solution.  Indeed, you already mentioned a slight concern about using a sessionid cookie.

One of the REALLY huge benefits that we've realized by using Blazor Server Side (?Server Interactive?) is that there is no SessionID cookie.  Unlike ASP.Net WebForms - each browser tab gets it's own independent Circuit.  This allows each Browser Tab to act completely independently of other tabs.  Now it is possible to have:
1.  One user open multiple tabs having independent app logins going at the same time.  (user can login to view Item num 1 and on another tab login to view Item num 2)
2. Two users sitting at the same PC interact with the app at the same time in their own tab (e.g. trainer/trainee)
3. Other use cases I won't discuss here

As you know cookies are a cross-tab resource.  I recently verified this by having one user login with multiple tabs while using the Azure Load Balancer solution with session affinity turned on.  The Azure load balancer puts a cookie on the browser and all the tabs pointing to the site get sent to the same server.

So my long winded point is that requiring the use of a session cookie would cause the (perhaps unintended) side effect of all browser tabs sharing the same session.  With your idea still in draft, I figured I'd put this bug in your ear as this is maybe some unique use of blazor server side that you hadn't thought of.

**Comment by rockfordlhotka** [comment]
Thanks @swegele!

There is a cookie scope that is per-tab, and that's what I'm using in the `Blazor8State` project, so it doesn't interfere with having different user sessions by having different tabs.

I've explored some other ideas around maintaining a session id value that is consistent in server-rendered, server-interactive, and wasm-interactive, but a cookie appears to be the only answer.

**Reply by swegele** [reply]
I'm not aware of this:

> There is a cookie scope that is per-tab

Do you have a link?  I wasn't able to find anything...is this new?

e.g.
https://github.com/dotnet/aspnetcore/issues/22167#issuecomment-633140937

**Reply by rockfordlhotka** [reply]
It appears to be the default scope - I didn't set a scope in my code. The scope name (in debug tools) shows as `Session`, and in my (limited) testing, I've seen that each tab gets its own cookie, and that the cookie goes away as soon as the tab/browser is closed.

**Reply by swegele** [reply]
I just wrote a quick empty blazor server app to test this ([inspired from here](https://stackoverflow.com/a/64687687)).
I used the Empty Blazor Server app template.  Made the index razor page have 1 button which simply writes a cookie via jsinterop.

Step 1 - Run the app.  Click the "Write" button. Code writes a cookie with DateTime.Now as the value.  
Step 2 - Open F12 dev tools: Application - see the cookie
Step 3 - Open a new tab and copy URL from first tab and enter (do not click any button)
Step 4 - Open F12 dev tools: Application - see the cookie is present from 1st tab in this tab too.

You can see the same cookie in both tabs.  
It is my understanding that cookies are shared for the same site/domain/thingy 
So if your session id is stored in a cookie...then by definition the other tabs (even thought they are connected via separate web socket circuits) will now share session via the cookie value.

![image](https://github.com/MarimerLLC/csla/assets/6868671/23874ccd-818d-4e06-8586-b1876a79d7e5)

That being said....I believe you could keep tab's session separate if you stored your session id in the browser's session storage.  
In BSS this is quite easy using Chris Sainty's [Blazored Session Storage](https://github.com/Blazored/SessionStorage) package.

**Reply by rockfordlhotka** [reply]
Hmm, maybe my testing was faulty.

I don't think the browser session storage will work for server-side code? At least not for server-rendered pages where the only thing on the client is html/css.

**Reply by swegele** [reply]
I'm happy to report that **both** local and session storage work great with the server-side code :-)  (which is just friggin cool that the blazor peeps figured that out).  We use it in our blazor server app and the sainty library makes it easy.

sorry i realized my above message made me sound unsure "I believe...".  What I meant to say was "I think that will work for your session id value too as I already know it works for storing other things in blazor server".

**Reply by rockfordlhotka** [reply]
When you say "server-side code" does that include server-rendered (aspnetcore) as well as server-interactive (Blazor server)?

**Reply by swegele** [reply]
Good question, which makes me realize how confusing things are going to get as now we have two separate server side technologies!  Ugh.  As if two types of Blazor wasn't enough...now we have 3!  
 
I have not tried using local/session storage with the new blazor server .Net 8 thing.  
I've only used it in the "old" Blazor Server things (which uses the SignalR circuit).

**Reply by rockfordlhotka** [reply]
That's the point of my blog posts - how to maintain some consistency between all THREE render models as they switch automatically (and dynamically) from page to page, or even as the user sits on one page!

Very magical and awesome, but with some real side-effects!

**Reply by swegele** [reply]
If writing/reading from there doesn't work for the new Blazor Server (non-signalR thing) then I wonder if you could make it adapt in situ-

The blazor server side (with signalR) and Blazor WASM could store the session id value in session storge...otherwise store it in a cookie?

**Reply by rockfordlhotka** [reply]
I think that's something the app developer would need to choose up front, because only they will know if they are including any server-rendered pages in their app.

In the (new) default template for a Blazor App you get all three. `Home` is server-rendered, as is the weather page (plus streamed content). The `Counter` page _starts_ as server-interactive and switches to wasm-interactive - potentially while the user is sitting on the page!
